File: recipes/claude/24-configure-stdio-and-sse-mcp-transports/typescript/src/server_sse.ts
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { SSEServerTransport } from "@modelcontextprotocol/sdk/server/sse.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";
import express, { Request, Response } from "express";

// Initialize core MCP Server instance
const server = new Server(
  {
    name: "ts-transport-showcase-server",
    version: "1.0.0",
  },
  {
    capabilities: {
      tools: {},
    },
  }
);

// Register tool listing handler
server.setRequestHandler(ListToolsRequestSchema, async () => {
  return {
    tools: [
      {
        name: "health_check",
        description: "Returns process uptime, memory consumption, and transport runtime statistics.",
        inputSchema: {
          type: "object",
          properties: {
            detailed: {
              type: "boolean",
              description: "Include memory layout and node environment stats.",
            },
          },
        },
      },
    ],
  };
});

// Register tool execution handler
server.setRequestHandler(CallToolRequestSchema, async (request) => {
  if (request.params.name !== "health_check") {
    throw new Error(`Tool not found: ${request.params.name}`);
  }

  const detailed = Boolean(request.params.arguments?.detailed);
  const memoryUsage = process.memoryUsage();

  const diagnosticData = {
    status: "healthy",
    uptimeSeconds: Math.floor(process.uptime()),
    pid: process.pid,
    nodeVersion: process.version,
    memory: detailed
      ? {
          heapUsedMb: (memoryUsage.heapUsed / 1024 / 1024).toFixed(2),
          heapTotalMb: (memoryUsage.heapTotal / 1024 / 1024).toFixed(2),
          rssMb: (memoryUsage.rss / 1024 / 1024).toFixed(2),
        }
      : undefined,
  };

  return {
    content: [
      {
        type: "text",
        text: JSON.stringify(diagnosticData, null, 2),
      },
    ],
  };
});

async function runStdio(): Promise<void> {
  console.error("[MCP] Initializing Stdio transport...");
  const transport = new StdioServerTransport();
  await server.connect(transport);
  console.error("[MCP] Stdio transport successfully connected.");
}

function runSSE(host: string, port: number): void {
  const app = express();
  let sseTransport: SSEServerTransport | null = null;

  // Endpoint 1: SSE Handshake and persistent stream
  app.get("/sse", async (req: Request, res: Response) => {
    console.error("[MCP] Inbound SSE connection initialized from", req.ip);
    
    // Construct transport targeting the downstream message POST path
    sseTransport = new SSEServerTransport("/messages", res);
    
    // Connect server to transport
    await server.connect(sseTransport);

    req.on("close", () => {
      console.error("[MCP] Client terminated SSE connection.");
    });
  });

  // Endpoint 2: Client JSON-RPC dispatch endpoint
  app.post("/messages", async (req: Request, res: Response) => {
    if (!sseTransport) {
      res.status(400).json({ error: "SSE transport stream not active. Initialize /sse first." });
      return;
    }
    console.error("[MCP] Processing inbound JSON-RPC message payload");
    await sseTransport.handlePostMessage(req, res);
  });

  app.listen(port, host, () => {
    console.error(`[MCP] SSE Transport Server listening at http://${host}:${port}`);
    console.error(`[MCP] SSE Stream: http://${host}:${port}/sse`);
    console.error(`[MCP] Messages:   http://${host}:${port}/messages`);
  });
}

// Bootstrap selector based on CLI arguments
const transportMode = process.argv.includes("--transport")
  ? process.argv[process.argv.indexOf("--transport") + 1]
  : process.env.MCP_TRANSPORT || "sse";

const host = process.env.MCP_HOST || "127.0.0.1";
const port = parseInt(process.env.MCP_PORT || "8000", 10);

if (transportMode === "stdio") {
  runStdio().catch((err) => {
    console.error("[MCP Fatal] Stdio transport failure:", err);
    process.exit(1);
  });
} else {
  runSSE(host, port);
}

File: recipes/claude/24-configure-stdio-and-sse-mcp-transports/python/server_sse.py
"""
Dual-Transport Model Context Protocol (MCP) Server in Python.
Supports standard I/O (stdio) transport and Server-Sent Events (SSE) over Starlette/Uvicorn.
"""

import argparse
import asyncio
import logging
import os
import sys
from typing import Any

import mcp.server.stdio
from mcp.server import NotificationOptions, Server
from mcp.server.models import InitializationOptions
from mcp.server.sse import SseServerTransport
import mcp.types as types
from starlette.applications import Starlette
from starlette.routing import Route
import uvicorn

# Configure structured logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
    handlers=[logging.StreamHandler(sys.stderr)],
)
logger = logging.getLogger("mcp-transport-server")

# Instantiate server instance with capability declarations
server = Server("transport-showcase-server")


@server.list_tools()
async def handle_list_tools() -> list[types.Tool]:
    """Declare available tool definitions."""
    return [
        types.Tool(
            name="system_diagnostic",
            description="Returns system runtime statistics, transport mode, and process health.",
            inputSchema={
                "type": "object",
                "properties": {
                    "verbose": {
                        "type": "boolean",
                        "description": "Whether to return verbose diagnostic attributes.",
                        "default": False,
                    }
                },
            },
        )
    ]


@server.call_tool()
async def handle_call_tool(
    name: str, arguments: dict[str, Any] | None
) -> list[types.TextContent]:
    """Handle incoming tool invocation requests."""
    if name != "system_diagnostic":
        raise ValueError(f"Unknown tool requested: {name}")

    args = arguments or {}
    verbose = args.get("verbose", False)

    diagnostic_payload = {
        "status": "operational",
        "transport": os.environ.get("MCP_TRANSPORT", "active"),
        "pid": os.getpid(),
        "python_version": sys.version.split()[0],
        "verbose_mode": verbose,
    }

    return [
        types.TextContent(
            type="text",
            text=f"System Diagnostic Result:\n{diagnostic_payload}",
        )
    ]


async def run_stdio() -> None:
    """Run server over standard input and output channels."""
    logger.info("Initializing Stdio transport runner...")
    async with mcp.server.stdio.stdio_server() as (read_stream, write_stream):
        await server.run(
            read_stream,
            write_stream,
            InitializationOptions(
                server_name="transport-showcase-server",
                server_version="1.0.0",
                capabilities=server.get_capabilities(
                    notification_options=NotificationOptions(),
                    experimental_capabilities={},
                ),
            ),
        )


def create_sse_application() -> Starlette:
    """Construct Starlette application for SSE transport endpoints."""
    sse = SseServerTransport("/messages")

    async def handle_sse(request: Any) -> Any:
        async with sse.connect_sse(
            request.scope, request.receive, request._send
        ) as streams:
            await server.run(
                streams[0],
                streams[1],
                InitializationOptions(
                    server_name="transport-showcase-server",
                    server_version="1.0.0",
                    capabilities=server.get_capabilities(
                        notification_options=NotificationOptions(),
                        experimental_capabilities={},
                    ),
                ),
            )

    async def handle_messages(request: Any) -> Any:
        await sse.handle_post_message(request.scope, request.receive, request._send)

    routes = [
        Route("/sse", endpoint=handle_sse),
        Route("/messages", endpoint=handle_messages, methods=["POST"]),
    ]

    return Starlette(debug=True, routes=routes)


def main() -> None:
    """Parse runtime arguments and launch chosen transport listener."""
    parser = argparse.ArgumentParser(
        description="Dual-Transport Model Context Protocol (MCP) Server"
    )
    parser.add_argument(
        "--transport",
        choices=["stdio", "sse"],
        default=os.environ.get("MCP_TRANSPORT", "sse"),
        help="Transport mechanism to bind (stdio or sse).",
    )
    parser.add_argument(
        "--host",
        default=os.environ.get("MCP_HOST", "127.0.0.1"),
        help="Host interface to bind HTTP listener.",
    )
    parser.add_argument(
        "--port",
        type=int,
        default=int(os.environ.get("MCP_PORT", "8000")),
        help="Port number for HTTP SSE listener.",
    )

    args = parser.parse_args()

    if args.transport == "stdio":
        logger.info("Starting MCP server in stdio mode...")
        asyncio.run(run_stdio())
    elif args.transport == "sse":
        logger.info("Starting MCP server in SSE mode on %s:%s...", args.host, args.port)
        app = create_sse_application()
        uvicorn.run(app, host=args.host, port=args.port, log_level="info")


if __name__ == "__main__":
    main()
